### ISSUE TITLE
Possible nil reference panic when proto.Size called in init() func

### ISSUE THREAD

**Issue by jeffreystoke** [golang/protobuf#1138]
**What version of protobuf and what language are you using?**
Version: `v1.4.2`

**What did you do?**

Tried to switch from gogoproto to golang/protobuf, but when I call `proto.Size` (or any function related to message encoding inside the proto package) in `init()` func in my package, there is a chance to cause nil reference panic, in my case if the filename starts with `a`, `b` or `c` will cause this error, it seems the proto package has some expectations for the `init()` call sequence, so I tried this:

filename: foo.proto

```protobuf
syntax = "proto3";

package main;

option go_package = "main";

message Foo {
  string bar = 1;
}
```

generated `foo.pb.go` file with `protoc --go_out=. --go_opt=paths=source_relative ./foo.proto`

filename: cmd.go (any filename will do)

```go
package main

import "google.golang.org/protobuf/proto"

var size int

func init() {
	size = proto.Size(&Foo{})
}

func main() {
	println(size)
}
```

then build with `go build -o main ./cmd.go ./foo.pb.go`

the example above can reproduce this issue every time

**What did you expect to see?**

The compiled program prints `0` and exit

**What did you see instead?**

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x98 pc=0x119cc59]

goroutine 1 [running]:
google.golang.org/protobuf/internal/impl.(*MessageInfo).initOnce(0xc00011a640)
	/go/pkg/mod/google.golang.org/protobuf@v1.23.0/internal/impl/message.go:84 +0x99
google.golang.org/protobuf/internal/impl.(*MessageInfo).init(0xc00011a640)
	/go/pkg/mod/google.golang.org/protobuf@v1.23.0/internal/impl/message.go:72 +0x3c
google.golang.org/protobuf/internal/impl.(*messageState).ProtoMethods(0xc00010e240, 0xc00010e240)
	/go/pkg/mod/google.golang.org/protobuf@v1.23.0/internal/impl/message_reflect_gen.go:30 +0x38
google.golang.org/protobuf/proto.protoMethods(...)
	/go/pkg/mod/google.golang.org/protobuf@v1.23.0/proto/proto_methods.go:18
google.golang.org/protobuf/proto.sizeMessage(0x126a580, 0xc00010e240, 0xc00010e240)
	/go/pkg/mod/google.golang.org/protobuf@v1.23.0/proto/size.go:30 +0x35
google.golang.org/protobuf/proto.MarshalOptions.Size(0x1000000, 0x12651e0, 0xc00010e240, 0xc00011a640)
	/go/pkg/mod/google.golang.org/protobuf@v1.23.0/proto/size.go:26 +0x4e
google.golang.org/protobuf/proto.Size(...)
	/go/pkg/mod/google.golang.org/protobuf@v1.23.0/proto/size.go:16
main.init.0()
	/play/protobuf-reproduce/cmd.go:8 +0x53
```

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

**Anything else we should know about your project / environment?**

go version: `go version go1.14.3 darwin/amd64`

**Comment by puellanivis** [comment]
TLDR: This is actually working as intended. That’s why best practice is to not mix your `.pb.go` files with the code that is using them. Especially in cases of `init()` functions, where initialization order is determined by lexigraphical order of the filenames. (You’ll find a `d.go`, `e.go` and `f.go` will also crash, but not a `fop.go`.)

Long explanation: the problem is that your proto has yet to be registered. Of note is this, from [the language specification](https://golang.org/ref/spec#Package_initialization) about package initialization:

> To ensure reproducible initialization behavior, build systems are encouraged to present multiple files belonging to the same package in lexical file name order to a compiler.

So, if you put your `.pb.go` file in the same directory as a a `func init()` that references it, you will run into a problem where the lexical ordering of the files determines if the protobuf has been registered or not when you start into your own `func init()` functions. This is obviously undesirable.

The wide recommendation of best practice is to generate your `.pb.go` files into a separate package than can then be imported by the code that uses them. The ordering of initialization in this case will guarantee that all of the subpackage’s `init()` functions (i.e. the proto registrations) have been called before starting in on the package’s initialization, meaning your proto will be guaranteed to have been properly registered at that time.

**Comment by jeffreystoke** [comment]
@puellanivis Thanks for your detailed message, that's very helpful, I'm aware of the `init()` problem in go, and can understand the best practice, that's good, and should be applied to every package (maybe including this one).

just like what I said, I was migrating the code from gogoproto (which worked well with this use case), I guess we can do something like explicit initialisation in proto library since the init() function in proto package will call `initOnce` at last, not sure if that's viable in your opinion?

**Comment by dsnet** [comment]
Thanks @puellanivis, I agree with your assessment. I'm surprised how fast you're able to diagnose that :)

> I guess we can do something like explicit initialisation in proto library since the init() function in proto package will call initOnce at last, not sure if that's viable in your opinion?

I'm not sure what you mean by "explicit initialization".  Did you mean something like manually calling the `file_xxx_proto_init` function? In theory that would work, but it's almost certainly going to break again in the future. These functions are unexported for a reason.

The [compatibility document for Go protobufs](https://github.com/protocolbuffers/protobuf-go#compatibility) states:

> We reserve the right to add new declarations to generated Go packages of .proto files. This includes declared constants, variables, functions, types, fields in structs, and methods on types. This may break attempts at injecting additional code on top of what is generated by protoc-gen-go. Such practice is not supported by this project.

The compatibility statement disallows adding additional code alongside the generated one because it would be impossible for the protobuf maintainers to reason about what will be a compatible versus and incompatible change in the future.

**Comment by jeffreystoke** [comment]
Thanks @dsnet , it's not a big issue, we can just follow the best practice anyway (why not), just got a little bit surprised  when it compiles but not able to run due to `init()` :)

> I'm not sure what you mean by "explicit initialization"

I mean maybe we can add a global exported function `proto.Init()` or something like that, so someone can call it first before running code in `init()`

**Comment by dsnet** [comment]
> I mean maybe we can add a global exported function proto.Init()

Given that our compatibility policy calls out non-support for injected code within the generated package, I don't think this functionality provides any benefit. Assuming users are importing the generated proto packages, then the Go language guarantees that the init functions for a depended upon package must occur before any possible uses of that package.

**Comment by jeffreystoke** [comment]
Thanks @dsnet and @puellanivis , I get it, closing this issue.
